### PR TITLE
fix resnet50_quantized_qat and mobilenet_v2_quantized_qat <> functionalization

### PR DESCRIPTION
**Pull request by bdhirsh**
This won't actually fix the issue until we make FakeTensor always-on for AOTAutograd.

I confirmed with the following benchmark (with `normalize_ir=False` and `use_functionalize=True`) in the dynamo/functorch config (run inside the `torch dynamo` repo):
```
terminal...$ python benchmarks/torchbench.py --training --devices=cuda --accuracy-aot-nop --generate-aot-autograd-stats --use-eval-mode --isolate --only=mobilenet_v2_quantized_qat

cuda train mobilenet_v2_quantized_qat         0.967x p=0.00

terminal...$ python benchmarks/torchbench.py --training --devices=cuda --accuracy-aot-nop --generate-aot-autograd-stats --use-eval-mode --isolate --only=resnet50_quantized_qat

cuda train resnet50_quantized_qat             0.943x p=0.00
```

I explained a bit more in the comment: quantized models use a running-mean style op, `fused_moving_avg_obs_fake_quant`, that takes in the running min/max stored on the module and mutates them, potentially resizing them.

That causes `AOTAutograd` to complain: AOTAutograd first takes views of the inputs (using `.detach().requires_grad_(grad)`), and plumbs them through the function to figure out what output to trace the backward with. These new inputs now have `TensorImpl::allow_tensor_metadata_change_ = false`, which causes the op to fail when it tries to resize the running counter variables. Once we're always using fake tensors, we shouldn't need to use `.detach().requires_grad_()` anymore (since we already have fresh fake tensors to trace with).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83542
* #83457
* __->__ #83339

### CODE REVIEW

**Comment by bdhirsh** [comment]
One thing I was thinking about was: technically wouldn't the `detach().requires_grad(grad)` pattern be ok, if we functionalized the function first (so we use functionalization earlier in AOTAutograd).

I think this is unnecessary once we start using fake tensors though, so it didn't seem worth trying to get that to work.

File: functorch/functorch/_src/aot_autograd.py [hunk 1]
```diff
@@ -197,11 +197,25 @@ def forward(ctx, *flat_tensor_args):
...
+                if config.use_fake_tensor:
+                    fake_mode = FakeTensorMode.push()
+                else:
+                    fake_mode = nullcontext()
+                    # The detach().requires_grad_() pattern can cause some subtle bugs.
+                    # These will be fixed once FakeTensor is always-on for AOTAutograd.
+                    #
+                    # For models that might resize their inputs, the input tensors
+                    # must have allow_tensor_metadata_change() set to true.
+                    # detach() returns a view tensor, but with that field set to false.
```
Comment (ezyang) [inline functorch/functorch/_src/aot_autograd.py:209]: why don't you just clone lol

**Review by ezyang — APPROVED** [review]

File: functorch/functorch/_src/aot_autograd.py (hunk 1 above)
Comment (bdhirsh) [reply functorch/functorch/_src/aot_autograd.py:209]: yeah, I just figured that might lead to other errors (OOM'ing during tracing), and we're hopefully going to kill this soon once FakeTensor mode is always on anyway
